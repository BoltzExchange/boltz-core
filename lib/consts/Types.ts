import { BIP32Interface } from 'bip32';
import { ECPairInterface } from 'ecpair';
import { TxOutput } from 'bitcoinjs-lib';
import { OutputType } from './Enums';

export type Error = {
  message: string;
  code: string;
};

export type ScriptElement = Buffer | number | string;

export type TransactionOutput = {
  txHash: Buffer;
  vout: number;
  type: OutputType;
} & TxOutput;

export type RefundDetails = TransactionOutput & {
  keys: ECPairInterface | BIP32Interface;
  redeemScript: Buffer;
};

export type ClaimDetails = RefundDetails & {
  preimage: Buffer;
};
