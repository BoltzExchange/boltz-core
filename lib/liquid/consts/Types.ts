import { Transaction, TxOutput } from 'liquidjs-lib';
import { RefundDetails } from '../../consts/Types';

export type LiquidRefundDetails = RefundDetails &
  TxOutput & {
    legacyTx?: Transaction;
    blindingPrivKey?: Buffer;
  };

export type LiquidClaimDetails = LiquidRefundDetails & {
  preimage: Buffer;
};
