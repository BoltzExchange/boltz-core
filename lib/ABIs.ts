import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
