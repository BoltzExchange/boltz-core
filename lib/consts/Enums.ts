export enum OutputType {
  Bech32 = 0,
  Compatibility = 1,
  Legacy = 2,
  Taproot = 3,
}
