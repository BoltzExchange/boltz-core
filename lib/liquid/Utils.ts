import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
