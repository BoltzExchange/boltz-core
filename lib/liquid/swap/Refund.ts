import { Transaction } from 'liquidjs-lib';
import { Network } from 'liquidjs-lib/src/networks';
import { getHexBuffer } from '../../Utils';
import Networks from '../consts/Networks';
import { LiquidClaimDetails, LiquidRefundDetails } from '../consts/Types';
import { constructClaimTransaction } from './Claim';

const dummyPreimage = getHexBuffer('0x00');

/**
 * Refund swaps
 *
 * @param utxos UTXOs that should be refunded
 * @param destinationScript the output script to which the funds should be sent
 * @param timeoutBlockHeight locktime of the transaction; only needed if the transaction is a refund
 * @param fee how many satoshis should be paid as fee
 * @param isRbf whether the transaction should signal full Replace-by-Fee
 * @param network network of the transaction
 * @param blindingKey blinding public key for the output; undefined if the output should not be blinded
 */
export const constructRefundTransaction = (
  utxos: LiquidRefundDetails[],
  destinationScript: Buffer,
  timeoutBlockHeight: number,
  fee: number,
  isRbf = true,
  network: Network = Networks.liquidMainnet,
  blindingKey?: Buffer,
): Transaction => {
  const claimUtxos: LiquidClaimDetails[] = utxos.map((utxo) => ({
    preimage: dummyPreimage,
    ...utxo,
  }));

  return constructClaimTransaction(
    claimUtxos,
    destinationScript,
    fee,
    isRbf,
    network,
    blindingKey,
    timeoutBlockHeight,
    true,
  );
};
