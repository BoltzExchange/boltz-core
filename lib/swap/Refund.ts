/**
 * This file is based on the repository github.com/submarineswaps/swaps-service created by Alex Bosworth
 */

import { getHexBuffer } from '../Utils';
import { constructClaimTransaction } from './Claim';
import { RefundDetails, ClaimDetails } from '../consts/Types';

const dummyPreimage = getHexBuffer('0x00');

/**
 * Refund swaps
 *
 * @param utxos UTXOs that should be refunded
 * @param destinationScript the output script to which the funds should be sent
 * @param timeoutBlockHeight locktime of the transaction
 * @param feePerByte how many satoshis per vbyte should be paid as fee
 *
 * @returns refund transaction
 */
export const constructRefundTransaction = (utxos: RefundDetails[], destinationScript: Buffer, timeoutBlockHeight: number, feePerByte: number) => {
  const claimUtxos: ClaimDetails[] = [];

  utxos.forEach((utxo) => {
    claimUtxos.push({
      preimage: dummyPreimage,
      ...utxo,
    });
  });

  return constructClaimTransaction(
    claimUtxos,
    destinationScript,
    feePerByte,
    true,
    timeoutBlockHeight,
  );
};
