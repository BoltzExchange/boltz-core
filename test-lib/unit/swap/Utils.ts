import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
