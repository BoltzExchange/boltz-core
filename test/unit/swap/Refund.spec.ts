import BIP32Factory from 'bip32';
import * as ecc from 'tiny-secp256k1';
import { getHexBuffer } from '../../../lib/Utils';
import { OutputType } from '../../../lib/consts/Enums';
import { RefundDetails } from '../../../lib/consts/Types';
import { constructRefundTransaction } from '../../../lib/swap/Refund';

const bip32 = BIP32Factory(ecc);

describe('Refund', () => {
  const utxo = {
    txHash: getHexBuffer(
      '285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d754',
    ),
    vout: 0,
    value: 2000,

    keys: bip32.fromBase58(
      'xprv9xgxR6htMdXUXGipynZp1janNrWNYJxaz2o4tH9fdtZqcF26BX5VB88GSM5KgZHWCyAyb8FZpQik2UET84CHfGWXFMG5zWWjmtDMgqYuo19',
    ),
    redeemScript: getHexBuffer(
      'a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac',
    ),
  };

  const refundDetails = [
    {
      ...utxo,
      type: OutputType.Bech32,
      script: getHexBuffer(
        '00206f38b6ce82427d4df080a9833d06cc6c66ab816545c9fd4df50f9d1ca8430b9e',
      ),
    },
    {
      ...utxo,
      type: OutputType.Legacy,
      script: getHexBuffer('a9143cdeb56e328a10d3bfe107fd5a16bd73871adb8d87'),
    },
    {
      ...utxo,
      type: OutputType.Compatibility,
      script: getHexBuffer('a9148f439aff651860bdb28c66500c6e958cfbe7a69387'),
    },
  ];

  const refundDetailsMap = new Map<OutputType, RefundDetails>(
    refundDetails.map((entry) => [entry.type, entry]),
  );

  const testRefund = (utxos: RefundDetails[], fee: number) => {
    return constructRefundTransaction(
      utxos,
      getHexBuffer('00140000000000000000000000000000000000000000'),
      11,
      fee,
    );
  };

  test.each`
    type                        | fee    | name
    ${OutputType.Bech32}        | ${127} | ${'P2WSH'}
    ${OutputType.Compatibility} | ${162} | ${'P2SH nested P2WSH'}
    ${OutputType.Legacy}        | ${258} | ${'P2SH'}
  `(`should refund a $name swap`, async ({ type, fee }) => {
    expect(
      testRefund([refundDetailsMap.get(type)!], fee).toHex(),
    ).toMatchSnapshot();
  });

  test('should refund multiple swaps in one transaction', () => {
    expect(testRefund(refundDetails, 466).toHex()).toMatchSnapshot();
  });
});
