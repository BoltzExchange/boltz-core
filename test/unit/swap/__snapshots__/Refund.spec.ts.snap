// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Refund should refund a P2SH nested P2WSH swap 1`] = `"01000000000101285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000232200206f38b6ce82427d4df080a9833d06cc6c66ab816545c9fd4df50f9d1ca8430b9efdffffff012e070000000000001600140000000000000000000000000000000000000000034830450221008e25e3f4c50c093641aea31b9e66555e14a8984493e89277beac61a91cce1bd902205743f99091eb45fc2db8c97ae5b69cd1ed04188294c3055cbdf0118ee472ee5f010064a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0b000000"`;

exports[`Refund should refund a P2SH swap 1`] = `"0100000001285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000af473044022001c3dbb2ece8ccba1e523b4e47cceaa1f0dae9b9e344af2f612604a387c41da602201b0ece06de76cc24820623a5eac7a00cf59fdc5b9d2cd8ee5dcdb5625ed4bd9101004c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368acfdffffff01ce0600000000000016001400000000000000000000000000000000000000000b000000"`;

exports[`Refund should refund a P2WSH swap 1`] = `"01000000000101285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000fdffffff015107000000000000160014000000000000000000000000000000000000000003473044022001ec0e4acead9cc62047403df7752b0760b9f5fd8e50cbc9e363e698f00d6bc502207d132feddf67cf9dd800d6591920fb116b9b46f8080c53357344c7a6cf75fb83010064a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0b000000"`;

exports[`Refund should refund multiple swaps in one transaction 1`] = `"01000000000103285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d7540000000000fdffffff285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000af47304402201299de809dc2c67408c50940e7e35dfc4755b2ae0c152b7214f10ee308ad258802203ccc8d517504c587d522d281b587faf27a3d9e952fa7aae54abb6f22eb17c5d101004c64a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368acfdffffff285d227e2823c679c224b4d562a9b5b5b7b927badd483df9f4225c6fc761d75400000000232200206f38b6ce82427d4df080a9833d06cc6c66ab816545c9fd4df50f9d1ca8430b9efdffffff019e1500000000000016001400000000000000000000000000000000000000000347304402206054593b7a7ae1406ccdbcf42c6cf2a84d9e37d62a1fbb3eeee4cd249c70e834022054c92abe3d7793dabd9676b1e49bbe80e7137759874ec9e0de75ae2bb942edf2010064a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac000347304402206054593b7a7ae1406ccdbcf42c6cf2a84d9e37d62a1fbb3eeee4cd249c70e834022054c92abe3d7793dabd9676b1e49bbe80e7137759874ec9e0de75ae2bb942edf2010064a914a0738c92fde6361f09d28950c7bd0d2bf32b34be87632103be4a251dae719d565ce1d6a7a5787df99fc1ecc1f6e847567981a686f32abce167027802b1752103f7877d4ae985bb30b6f150ad6b6b9935c342432beed1a4781347b169c1e2417368ac0b000000"`;
